doc/subsystems/shell/shell.rst

```diff
@@ -0,0 +1,294 @@
+Characters within parentheses are treated as one argument. Each command or
+subcommand without a handler can be treated as an argument, or it can still
+have a subcommand with a handler and there is no problem for the user
+to execute it.
+
+Built-in commands
+=================
+
+* ``clear`` - Clears the screen.
+* ``shell`` - Useful shell-related commands like:
+	* ``echo`` - Toggles shell echo.
+        * ``colors`` - Toggles colored syntax. This might be helpful in case of Bluetooth shell to limit the amount of transferred bytes.
+        * ``stats`` - Shows shell statistics.
+* ``history`` - Shows the recently entered commands.
+* ``resize`` - Must be executed when terminal width is different than 80 characters or after each change of terminal width. It ensures proper multiline text display and Left/Right/End/Home buttons handling. It can be also called with a parameter (subcommand): ``resize default``. When run, the shell will send terminal width = 80 to the terminal and assume successful delivery. Currently ``resize`` command works only with UART flow control switched on.
+
+Commands help
+=============
+
+Every user-defined command, subcommand, or option can have its own help
+description. The help for commands and subcommands can be created with
+respective macros: :c:macro:`SHELL_CMD_REGISTER`, :c:macro:`SHELL_CMD`.
+In addition, you can define options for commands or subcommands using the
+macro :c:macro:`SHELL_OPT`. By default, each and every command or subcommand
+has these two options implemented: ``-h`` and ``--help``.
+
+In order to add help functionality to a command or subcommand, you must
+implement the help handler by either calling :cpp:func:`shell_cmd_precheck`
+or pair of functions :cpp:func:`shell_help_requested` and
+:cpp:func:`shell_help_print`. The former is more convinient as it also
+checks for valid arguments count.
+
+
+The following code is an example of how commands can be processed:
+
+.. code-block:: c
+
+	static void cmd_resize(const struct shell *shell, size_t argc,
+				char **argv)
+	{
+		int err;
+
+		if (!IS_ENABLED(CONFIG_SHELL_CMDS_RESIZE)) {
+			shell_fprintf(shell, SHELL_ERROR,
+					"Command not supported.\r\n");
+			return;
+		}
+
+		/* function responsible for printing command help */
+		if (!shell_cmd_precheck(shell, (argc <= 2), NULL, 0)) {
+			return;
+		}
+
+		if (argc != 1) {
+			shell_fprintf(shell, SHELL_ERROR, "%s:%s%s\r\n",
+				      argv[0],
+				      SHELL_MSG_UNKNOWN_PARAMETER, argv[1]);
+			return;
+		}
+
+		err = terminal_size_get(shell);
+		if (err != 0) {
+			shell->ctx->vt100_ctx.cons.terminal_wid =
+					SHELL_DEFAULT_TERMINAL_WIDTH;
+			shell->ctx->vt100_ctx.cons.terminal_hei =
+					SHELL_DEFAULT_TERMINAL_HEIGHT;
+			shell_fprintf(shell, SHELL_WARNING,
+				      "No response from the terminal, assumed"
+					" 80x24 screen size\r\n");
+		}
+	}
+
+It is recommended to use subcommands. Options apply mainly in the case when an
```

Comment:
Better would be:

    When possible, use subcommands instead of options.  Options apply  ...